src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1505 @@
+                                    // Equals without string allocation
+                                    return name.Length == memberName.Length + 4 &&
+                                           name.StartsWith("get_") && memberName.IndexOf(name, 4, StringComparison.Ordinal) == 4;
+                                }
+                                );
+                        foreach (var indexer in indexers)
+                        {
+                            yield return new PSTypeName(indexer.ReturnType);
+                        }
+                    }
+                }
+
+                // Inferred type of target wasn't indexable.  Assume (perhaps incorrectly)
+                // that it came from OutputType and that more than one object was returned
+                // and that we're indexing because of that, in which case, OutputType really
+                // is the inferred type.
+                yield return psType;
+            }
+        }
+
+        private IEnumerable<PSTypeName> GetInferredTypeFromScriptBlockParameter(AstParameterArgumentPair argument)
+        {
+            var argumentPair = argument as AstPair;
+            var scriptBlockExpressionAst = argumentPair?.Argument as ScriptBlockExpressionAst;
+            if (scriptBlockExpressionAst == null) yield break;
+            foreach (var type in InferTypes(scriptBlockExpressionAst.ScriptBlock))
+            {
+                yield return type;
+            }
+        }
+
+        object ICustomAstVisitor2.VisitTypeDefinition(TypeDefinitionAst typeDefinitionAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor2.VisitPropertyMember(PropertyMemberAst propertyMemberAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor2.VisitFunctionMember(FunctionMemberAst functionMemberAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor2.VisitBaseCtorInvokeMemberExpression(BaseCtorInvokeMemberExpressionAst baseCtorInvokeMemberExpressionAst)
+        {
+            return ((ICustomAstVisitor)this).VisitInvokeMemberExpression(baseCtorInvokeMemberExpressionAst);
+        }
+
+        object ICustomAstVisitor2.VisitUsingStatement(UsingStatementAst usingStatement)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor2.VisitConfigurationDefinition(ConfigurationDefinitionAst configurationDefinitionAst)
+        {
+            return configurationDefinitionAst.Body.Accept(this);
+        }
+
+        object ICustomAstVisitor2.VisitDynamicKeywordStatement(DynamicKeywordStatementAst dynamicKeywordAst)
+        {
+            // TODO: What is the right InferredType for the AST
+            return dynamicKeywordAst.CommandElements[0].Accept(this);
+        }
+    }
+}
```

Comment:
Now that we have code coverage data, it would be great to get some great coverage on this new file.